### ISSUE TITLE
ACT

### ISSUE THREAD

**Issue by wangyu-ustc** [wangyu-ustc/Denoising#21]
# for i, layer in enumerate(self.layers):
        #     mems_i = None if mems is None else mems[i]
        #
        #     if self.act_params is None:
        #         core_out, kv_mems, qkv_state = layer(core_out, pos_emb, self.r_w_bias, self.r_r_bias,
        #                                              dec_attn_mask=dec_attn_mask, mems=mems_i,
        #                                              running=last_layer_still_running, qkv_state=qkv_state)
        #     else:
        #         self.use_prev_state = self.act_params.get("use_prev_state", True)
        #         self.stop_type = self.act_params.get('stop_type', 'halt')
        #         if not (halting_probability < self.threshold).any():
        #             break
        #         p = self.sigmoid(self.state_proj(core_out))  # p: qlen, bsz, 1
        #
        #         if self.act_params['act_type'] == 'global':
        #             p = torch.mean(p, dim=0)
        #             p = torch.squeeze(p, dim=-1)  # bsz
        #         else:
        #             p = torch.squeeze(p, dim=-1)  # qlen, bsz
        #
        #         still_running = (halting_probability < 1.0).to(torch.float)
        #         new_halted = ((halting_probability + p * still_running) > self.threshold).to(
        #             torch.float) * still_running
        #         still_running = ((halting_probability + p * still_running) <= self.threshold).to(
        #             torch.float) * still_running
        #         # Add the halting probability for this step to the halting
        #         # probabilities for those input which haven't halted yet
        #         halting_probability = p * still_running + halting_probability
        #
        #         # Compute remainders for the inputs which halted at this step
        #         remainders = new_halted * (-halting_probability + 1) + remainders
        #
        #         # Add the remainders to those inputs which halted at this step
        #         halting_probability = new_halted * remainders + halting_probability
        #
        #         # Increment n_updates for all inputs which are still running
        #         last_layer_still_running = still_running + new_halted
        #         n_updates = last_layer_still_running + n_updates
        #
        #         # Compute the weight to be applied to the new state and output
        #         # 0 when the input has already halted
        #         # p when the input hasn't halted yet
        #         # the remainders when it halted this step
        #         update_weights = torch.unsqueeze(p * still_running + new_halted * remainders, -1)
        #         if self.act_params['act_type'] == 'global':
        #             update_weights = torch.unsqueeze(update_weights, dim=0)
        #
        #         # core_out: qlen, bsz, _
        #         new_core_out, kv_mems, qkv_state = layer(core_out, pos_emb, self.r_w_bias, self.r_r_bias,
        #                                   dec_attn_mask=dec_attn_mask, mems=mems_i,
        #                                   running=last_layer_still_running, qkv_state=qkv_state)
        #         if self.use_prev_state:
        #             previous_state = (new_core_out * update_weights) + \
        #                              (previous_state * (1 - update_weights))
        #             core_out = new_core_out
        #         else:
        #             core_out = new_core_out * update_weights + core_out * (1 - update_weights)
        #
        #     new_kv_mems.append(kv_mems)
        #
        # if mems is None:
        #     mems = [None] * self.n_layer
        #
        # new_mems = []
        # if self.is_decoding:
        #     new_mems = self._update_decode_mems(new_kv_mems, mems, [self.att_trunc_len] * len(self.layers))
        #
        # # calculating act_loss
        # if self.act_params is not None:
        #     n_layers = torch.mean(n_updates)
        #     act_loss += torch.mean(remainders) * self.act_params['act_loss_weight']
        #     if self.stop_type == 'halt' and self.use_prev_state:
        #         return previous_state, new_mems, (n_layers, act_loss)
        #     else:
        #         return core_out, new_mems, (n_layers, act_loss)
        # else:
        #     return core_out, new_mems, None
